lib/matplotlib/__init__.py

```diff
@@ -1296,17 +1278,35 @@ def _add_data_doc(docstring, replace_names):
             or replace_names is not None and len(replace_names) == 0):
         return docstring
     docstring = inspect.cleandoc(docstring)
-    repl = (
-        ("    every other argument can also be string ``s``, which is\n"
-         "    interpreted as ``data[s]`` (unless this raises an exception).")
-        if replace_names is None else
-        ("    the following arguments can also be string ``s``, which is\n"
-         "    interpreted as ``data[s]`` (unless this raises an exception):\n"
-         "    " + ", ".join(map("*{}*".format, replace_names))) + ".")
-    addendum = _DATA_DOC_APPENDIX.format(replaced=repl)
-    if _DATA_DOC_TITLE not in docstring:
-        addendum = _DATA_DOC_TITLE + addendum
-    return docstring + addendum
+
+    data_doc = ("""\
+    If given, all parameters also accept a string ``s``, which is
+    interpreted as ``data[s]`` (unless this raises an exception).
+
+    Objects passed as **data** must support item access (``data[s]``) and
+    membership test (``s in data``)."""
+                if replace_names is None else ("""\
+    If given, the following parameters also accept a string ``s``, which
+    is interpreted as ``data[s]`` (unless this raises an exception):
+
+    {names}
+
+    Objects passed as **data** must support item access (``data[s]``) and
+    membership test (``s in data``).""".format(
+            names=", ".join(map("*{}*".format, replace_names)))
+        )
+    )
+    # using string replacement instead of formatting has the advantages
+    # 1) simpler indent handling
+    # 2) prevent problems with formatting characters '{', '%' in the docstring
+    if _log.level <= logging.DEBUG:
+        # test_data_parameter_replacement() tests against these log messages
+        # make sure to keep message and test in sync
+        if "data : indexable object, optional" not in docstring:
+            _log.debug("data parameter docstring error: no data parameter")
+        if 'DATA_PARAMETER_PLACEHOLDER' not in docstring:
+            _log.debug("data parameter docstring error: missing placeholder")
```

Comment:
I think we should just throw (possibly via an assert) in that case?  this would be consistent with what _preprocess_data does for invalid replace_names.
(In any case, the exception should never reach end users, only CI...)

Reply:
I agree that the message should not reach end users. Particularly, I also don't want the checks to be run when the user imports matplotlib. I figured that tying execution to debug-logging is the easiest approach. Otherwise we'd need special flags or environment variables to check. I'm not worried that a user sees the message if he runs with debug logging. Debug logging outputs various other similarly technical messages.

Reply:
If you worry about performance, I think you can simply do:
```python
replaced = docstring.replace(..., ...)
assert replaced != docstring
return replaced
```
(without the `in` checks) as the string comparison should in practice be very fast as the strings won't even have the same length (which is checked first in the C-level `unicode_compare_eq`).  This may well be competitive with the two attribute accesses in `_log.level <= logging.DEBUG` :-) (well, tbh I haven't checked)

(Then you don't check that there's "data : indexable..." but I think that's fine, especially as CI also checks that the result is proper numpydoc.)

Reply:
I'm a bit hesitant with the assert. If something goes wrong, then matplotlib is not importable anymore.
And I don't think an ill-formatted docstring justifies terminating the library.

Reply:
But this should really not reach the end user; if somehow we miss this on CI then something is likely very wrong with our release process...
In any case I'm not blocking this if others agree with logging instead of asserting.

Reply:
I'm more concerned that someone outside our release process may tinker with the docstrings. E.g. someone could strip all docstrings for "performance reasons". While the specific case of a missing docstring is handled, it would break if all docstrings were replaced by "stripped". - Yes, that's an unlikely scenario. It appears to me that raising ever so slightly but unnecessarily increases the risk of breakage.

Reply:
OK, let's get this in.